src/sources/file/line_agg.rs

```diff
@@ -0,0 +1,519 @@
+use bytes::{Bytes, BytesMut};
+use futures::{Async, Poll, Stream};
+use regex::bytes::Regex;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, VecDeque};
+use std::time::Duration;
+use tokio::timer::DelayQueue;
+
+#[derive(Debug, Hash, Clone, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Mode {
+    /// All consecutive lines matching this pattern are included in the group.
+    /// The first line (the line that matched the start pattern) does not need
+    /// to match the `ContinueThrough` pattern.
+    /// This is useful in cases such as a Java stack trace, where some indicator
+    /// in the line (such as leading whitespace) indicates that it is an
+    /// extension of the preceeding line.
+    ContinueThrough,
+
+    /// All consecutive lines matching this pattern, plus one additional line,
+    /// are included in the group.
+    /// This is useful in cases where a log message ends with a continuation
+    /// marker, such as a backslash, indicating that the following line is part
+    /// of the same message.
+    ContinuePast,
+
+    /// All consecutive lines not matching this pattern are included in the
+    /// group.
+    /// This is useful where a log line contains a marker indicating that it
+    /// begins a new message.
+    HaltBefore,
+
+    /// All consecutive lines, up to and including the first line matching this
+    /// pattern, are included in the group.
+    /// This is useful where a log line ends with a termination marker, such as
+    /// a semicolon.
+    HaltWith,
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct Config {
+    /// Start pattern to look for as a beginning of the message.
+    pub start_pattern: Regex,
+    /// Condition pattern to look for. Exact behavior is configured via `mode`.
+    pub condition_pattern: Regex,
+    /// Mode of operation, specifies how the condition pattern is interpreted.
+    pub mode: Mode,
+    /// The maximum time to wait for the continuation. Once this timeout is
+    /// reached, the buffered message is guaraneed to be flushed, even if
+    /// incomplete.
+    pub timeout: Duration,
+}
+
+impl Config {
+    pub(super) fn for_legacy(marker: Regex, timeout_ms: u64) -> Self {
+        let start_pattern = marker;
+        let condition_pattern = start_pattern.clone();
+        let mode = Mode::HaltBefore;
+        let timeout = Duration::from_millis(timeout_ms);
+
+        Self {
+            start_pattern,
+            condition_pattern,
+            mode,
+            timeout,
+        }
+    }
+}
+
+/// Private type alias to be more expressive in the internal implementation.
+type Filename = String;
```

Comment:
I have two reservations about this. First, the most appropriate type for filenames would be `PathBuf` instead of `String`, which is already more obviously a filename. Second, I am wondering if the newtype pattern would be more appropriate here. Since they are used as keys to a `HashMap`, that implies extra code, though. What do you think @LucioFranco?

Reply:
If we change the underlying type, I like using the `PathBuf`. It’s better to update it at the whole file source though, and maybe make `LineAgg` generic over it. It would help making it more reusable, since we already have `event::Discriminant` that could also be used there.